File: src/cli/say.ts
import { defineCommand } from "citty";
import { errorMessage } from "../error-utils";
import {
  getEngineBinPath,
  isEngineInstalled,
  spawnEngineProcess,
  spawnStdioWithDebugFd,
} from "../engine";
import { installHint } from "../install-hint";
import { registerProcessTree } from "../process-tree";
import { log } from "../log";
import {
  say,
  SayError,
  SUPPORTED_SAMPLE_RATES,
  type SayFormat,
  type SayOptions,
} from "../synth";
import { artifactFromBytes, artifactFromFile, type StatsRecorder } from "../stats";
import { resolveSayVoice } from "../voice-routing";
import { diagnosticCharBucket, diagnosticSizeBucket } from "../diagnostic-events";
import { runCommandSession, type CommandOutcome, type CommandSession } from "./command-session";

async function resolveText(inline: string | undefined): Promise<string> {
  if (inline !== undefined && inline.length > 0) return inline;
  const chunks: Uint8Array[] = [];
  for await (const chunk of Bun.stdin.stream()) {
    chunks.push(chunk);
  }
  const total = chunks.reduce((n, c) => n + c.byteLength, 0);
  const merged = new Uint8Array(total);
  let offset = 0;
  for (const c of chunks) {
    merged.set(c, offset);
    offset += c.byteLength;
  }
  return new TextDecoder().decode(merged).trim();
}

export function shouldRejectMissingSayText(
  inlineText: string | undefined,
  stdinIsTty: boolean | undefined,
): boolean {
  return (inlineText === undefined || inlineText.length === 0) && stdinIsTty === true;
}

type Parsed<T> = { ok: true; value: T | undefined } | { ok: false; error: string };

function parseFiniteNumberFlag(name: string, value: unknown): Parsed<number> {
  if (value === undefined || value === null || value === false) return { ok: true, value: undefined };
  const raw = String(value).trim();
  const parsed = Number(raw);
  if (raw === "" || !Number.isFinite(parsed)) {
    return { ok: false, error: `${name} must be a finite number.` };
  }
  return { ok: true, value: parsed };
}

function parseRateFlag(value: unknown): Parsed<number> {
  const rate = parseFiniteNumberFlag("--rate", value);
  if (!rate.ok || rate.value === undefined) return rate;
  if (rate.value < 0.5 || rate.value > 2.0) {
    return { ok: false, error: "--rate must be between 0.5 and 2.0." };
  }
  return rate;
}

function parseBitrateFlag(value: unknown): Parsed<number> {
  const bitrate = parseFiniteNumberFlag("--bitrate", value);
  if (!bitrate.ok || bitrate.value === undefined) return bitrate;
  if (!Number.isInteger(bitrate.value) || bitrate.value <= 0) {
    return { ok: false, error: "--bitrate must be a positive integer." };
  }
  return bitrate;
}

function parseSampleRateFlag(value: unknown): Parsed<number> {
  const sampleRate = parseFiniteNumberFlag("--sample-rate", value);
  if (!sampleRate.ok || sampleRate.value === undefined) return sampleRate;
  if (
    !Number.isInteger(sampleRate.value) ||
    !SUPPORTED_SAMPLE_RATES.includes(sampleRate.value as (typeof SUPPORTED_SAMPLE_RATES)[number])
  ) {
    return {
      ok: false,
      error: `--sample-rate must be one of ${SUPPORTED_SAMPLE_RATES.join(", ")}.`,
    };
  }
  return sampleRate;
}

function parseFormatFlag(raw: unknown): Parsed<SayFormat> {
  if (typeof raw !== "string" || raw === "") return { ok: true, value: undefined };
  const lower = raw.toLowerCase();
  if (lower === "wav" || lower === "ogg-opus" || lower === "flac") return { ok: true, value: lower };
  if (lower === "opus" || lower === "ogg") return { ok: true, value: "ogg-opus" };
  return { ok: false, error: `unknown --format '${raw}'. supported: wav, ogg-opus, flac` };
}

function checkOpusOnlyFlags(
  bitrate: number | undefined,
  sampleRate: number | undefined,
  format: SayFormat | undefined,
  outArg: string | undefined,
): string | null {
  if (bitrate === undefined && sampleRate === undefined) return null;
  const outExt = outArg?.split(".").pop()?.toLowerCase();
  const impliesOpus = outExt !== undefined && ["ogg", "opus", "oga"].includes(outExt);
  // --bitrate / --sample-rate are Opus-only (wav/flac are lossless, no encoder knobs).
  const resolvesToOpus = format === "ogg-opus" || (format === undefined && impliesOpus);
  if (resolvesToOpus) return null;
  return "--bitrate and --sample-rate are only valid with --format ogg-opus";
}

export type SayFlagArgs = {
  format?: unknown;
  rate?: unknown;
  bitrate?: unknown;
  "sample-rate"?: unknown;
  out?: unknown;
};

export type ResolvedSayFlags =
  | {
      ok: true;
      format: SayFormat | undefined;
      rate: number | undefined;
      bitrate: number | undefined;
      sampleRate: number | undefined;
    }
  | { ok: false; error: string };

/** Validates the encoder flags before the engine is spawned: faster failure in scripts (the engine repeats them authoritatively). */
export function resolveSayFlags(args: SayFlagArgs): ResolvedSayFlags {
  const format = parseFormatFlag(args.format);
  if (!format.ok) return format;
  const rate = parseRateFlag(args.rate);
  if (!rate.ok) return rate;
  const bitrate = parseBitrateFlag(args.bitrate);
  if (!bitrate.ok) return bitrate;
  const sampleRate = parseSampleRateFlag(args["sample-rate"]);
  if (!sampleRate.ok) return sampleRate;

  const out = typeof args.out === "string" ? args.out : undefined;
  const opusOnlyError = checkOpusOnlyFlags(bitrate.value, sampleRate.value, format.value, out);
  if (opusOnlyError) return { ok: false, error: opusOnlyError };

  return {
    ok: true,
    format: format.value,
    rate: rate.value,
    bitrate: bitrate.value,
    sampleRate: sampleRate.value,
  };
}

type SayOpts = {
  text: string;
  voice: string | undefined;
  lang: string | undefined;
  out: string | undefined;
  rate: number | undefined;
  ssml: boolean;
  format: SayFormat | undefined;
  bitrate: number | undefined;
  sampleRate: number | undefined;
  noExpandAbbrev: boolean;
};

function recordOutputArtifact(
  stats: StatsRecorder,
  audio: Uint8Array,
  opts: SayOpts,
): { outputFormat: string; outputSizeBytes: number | null | undefined } {
  if (opts.out) {
    const outputArtifact = artifactFromFile(opts.out, "output_audio");
    if (outputArtifact) stats.recordArtifact(outputArtifact);
    return {
      outputFormat: outputArtifact?.format || opts.format || "auto",
      outputSizeBytes: outputArtifact?.sizeBytes,
    };
  }
  stats.recordArtifact(artifactFromBytes(audio.byteLength, "output_audio", opts.format ?? "wav"));
  return { outputFormat: opts.format ?? "wav", outputSizeBytes: audio.byteLength };
}

async function synthesizeAndEmit(
  opts: SayOpts,
  verbose: boolean,
  session: CommandSession,
): Promise<CommandOutcome> {
  const { stats } = session;
  try {
    const startedAt = performance.now();
    if (opts.out) {
      const voiceLabel = opts.voice ?? "default voice";
      log.status(`Synthesizing ${voiceLabel} -> ${opts.out}...`);
    }
    const audio = await stats.timeStage("tts", () => say(opts as SayOptions));
    const ttsTimeMs = Math.round(performance.now() - startedAt);
    if (opts.out) {
      log.status(`Saved ${opts.out} (${ttsTimeMs}ms)`);
    }
    // stderr — stdout may carry raw audio bytes when --out is omitted.
    if (verbose && !opts.out) log.status(`TTS time: ${ttsTimeMs}ms`);
    const { outputFormat, outputSizeBytes } = recordOutputArtifact(stats, audio, opts);
    if (!opts.out) process.stdout.write(audio);
    return {
      status: "success",
      itemCount: 1,
      finishFields: {
        durationMs: ttsTimeMs,
        hasOut: Boolean(opts.out),
        outputFormat,
        outputSizeBucket: diagnosticSizeBucket(outputSizeBytes),
      },
    };
  } catch (err) {
    const code = err instanceof SayError ? err.code : "E_INTERNAL";
    const exitCode = err instanceof SayError ? err.exitCode : 4;
    stats.recordError("tts", err, code);
    log.error(err instanceof SayError ? err.stderr.trim() || err.message : errorMessage(err));
    return {
      status: "failed",
      itemCount: 1,
      exitCode,
      finishFields: {
        errorKind: err instanceof SayError ? "say_error" : "error",
        exitCode,
        error_code: code,
      },
    };
  }
}

export const sayCommand = defineCommand({
  meta: {
    name: "say",
    description:
      "Synthesize speech from text (TTS). Writes audio to stdout (or --out file). Defaults to WAV; use --format ogg-opus for messenger-ready voice notes.",
  },
  args: {
    text: { type: "positional", required: false, description: "Text to speak (stdin if omitted)" },
    voice: { type: "string", description: "Voice id, e.g. en-am_michael" },
    lang: {
      type: "string",
      description:
        "Language code, e.g. en-us (default en-us; see docs/languages.md). With no --voice, routes to that language's default voice and skips text-language detection.",
    },
    out: { type: "string", description: "Write audio to file instead of stdout" },
    rate: { type: "string", description: "Speaking rate 0.5–2.0", default: "1.0" },
    "list-voices": { type: "boolean", description: "List installed voices and exit" },
    ssml: {
      type: "boolean",
      description: "Parse input as SSML (supports <speak>, <break>; strips unknown tags)",
    },
    format: {
      type: "string",
      description:
        "Output format: wav (default), ogg-opus (Telegram-ready voice note), or flac (lossless, plays in all browsers incl. Safari). Inferred from --out extension when omitted.",
    },
    bitrate: {
      type: "string",
      description: "Opus bitrate in bits/sec (e.g. 32000). Only with --format ogg-opus.",
    },
    "sample-rate": {
      type: "string",
      description:
        "Opus encoder sample rate (8000/12000/16000/24000/48000). Only with --format ogg-opus.",
    },
    "no-expand-abbrev": {
      type: "boolean",
      description:
        "Disable acronym auto-expansion on ru-vosk-* voices (ВОЗ → 'вэ о зэ') and English on ONNX engine builds " +
        "(FBI → 'ef bee eye'). No effect on macOS arm64 FluidAudio, macos-* or non-English voices, which spell " +
        "initialisms in their own G2P; the engine warns when the flag is ignored. " +
        "<say-as interpret-as='characters'> still works.",
    },
    verbose: {
      type: "boolean",
      description: "Log TTS synthesis time to stderr",
      default: false,
    },
    debug: {
      type: "boolean",
      description: "Trace engine subprocess calls on stderr (or KESHA_DEBUG=1)",
      default: false,
    },
  },
  async run({ args }) {
    if (args.debug) log.debugEnabled = true;
    if (args["list-voices"]) {
      if (!isEngineInstalled()) {
        log.error(`kesha-engine not installed. run: ${installHint()}`);
        process.exit(1);
      }
      // The engine prints the list directly — just relay its stdout + exit code.
      const proc = spawnEngineProcess(
        getEngineBinPath(),
        ["say", "--list-voices"],
        spawnStdioWithDebugFd(["inherit", "inherit", "inherit"]),
      );
      // Register so a Ctrl-C during a cold Engine load terminates it and exits 130/143 (#939);
      // dispose before the process.exit below so the registration never outlives the run.
      const tree = registerProcessTree(proc);
      let exitCode: number;
      try {
        exitCode = await proc.exited;
      } finally {
        tree.dispose();
      }
      process.exit(exitCode);
    }

    const flags = resolveSayFlags(args);
    if (!flags.ok) {
      log.error(flags.error);
      process.exit(2);
    }

    const inlineText = typeof args.text === "string" ? args.text : undefined;
    const stdinIsTty = (process.stdin as { isTTY?: boolean }).isTTY;
    if (shouldRejectMissingSayText(inlineText, stdinIsTty)) {
      log.error("kesha say requires text or piped stdin. Usage: kesha say <text>");
      process.exit(2);
    }
    const text = await resolveText(inlineText);
    const explicitVoice = typeof args.voice === "string" ? args.voice : undefined;
    const langHint = typeof args.lang === "string" ? args.lang : undefined;
    const voice = await resolveSayVoice(explicitVoice, langHint, text);

    const opts: SayOpts = {
      text,
      voice,
      lang: langHint,
      out: typeof args.out === "string" ? args.out : undefined,
      rate: flags.rate,
      ssml: Boolean(args.ssml),
      format: flags.format,
      bitrate: flags.bitrate,
      sampleRate: flags.sampleRate,
      noExpandAbbrev: Boolean(args["no-expand-abbrev"]),
    };

    const outcome = await runCommandSession(
      "say",
      {
        charBucket: diagnosticCharBucket(Array.from(text).length),
        hasInlineInput: inlineText !== undefined,
        hasVoice: explicitVoice !== undefined,
        autoVoice: explicitVoice === undefined && voice !== undefined,
        hasOut: typeof opts.out === "string",
        outputFormat: opts.format ?? "auto",
        ssml: opts.ssml,
        noExpandAbbrev: opts.noExpandAbbrev,
      },
      (session) => synthesizeAndEmit(opts, Boolean(args.verbose), session),
    );

    if (outcome.exitCode !== undefined) process.exit(outcome.exitCode);
  },
});

File: rust/src/cli/say.rs
use anyhow::Result;
use std::path::PathBuf;

use crate::protocol::events;
use crate::{models, say_loop, tts};

#[derive(clap::Args)]
pub struct SayArgs {
    /// Text to synthesize (omit to read from stdin)
    pub text: Option<String>,
    /// Voice id, e.g. `en-am_michael`
    #[arg(long)]
    pub voice: Option<String>,
    /// Override the voice's default BCP 47 language code, e.g. `en-gb`
    #[arg(long)]
    pub lang: Option<String>,
    /// Output file (default: stdout)
    #[arg(long)]
    pub out: Option<PathBuf>,
    /// Speaking rate (0.5–2.0)
    #[arg(long, default_value_t = 1.0)]
    pub rate: f32,
    /// List installed voices and exit
    #[arg(long)]
    pub list_voices: bool,
    /// Parse the input as SSML (supports <speak>, <break>; strips unknown tags).
    /// See issue #122 for the v1 tag matrix.
    #[arg(long)]
    pub ssml: bool,
    /// Output audio format. Defaults to `wav` (or inferred from `--out`
    /// extension when omitted). Supported: `wav`, `ogg-opus`. See #223.
    #[arg(long, value_name = "FORMAT")]
    pub format: Option<String>,
    /// Opus bitrate in bits/second (e.g. 16000, 32000, 64000). Only valid
    /// with `--format ogg-opus`. Default 32000 (Telegram-grade).
    #[arg(long, value_name = "BPS")]
    pub bitrate: Option<i32>,
    /// Encoder sample rate. Only valid with `--format ogg-opus`. Must be
    /// one of 8000/12000/16000/24000/48000. Default 24000.
    #[arg(long = "sample-rate", value_name = "HZ")]
    pub sample_rate: Option<u32>,
    /// Explicit model path (testing override)
    #[arg(long, hide = true)]
    pub model: Option<PathBuf>,
    /// Explicit voice embedding file (testing override)
    #[arg(long = "voice-file", hide = true)]
    pub voice_file: Option<PathBuf>,
    /// Long-lived loop: read newline-delimited JSON requests on stdin,
    /// reuse loaded engines across calls, write framed binary responses
    /// on stdout. See `docs/tts-stdin-loop.md`. Issue #213.
    #[arg(long = "stdin-loop", hide = true)]
    pub stdin_loop: bool,
    /// Disable acronym auto-expansion on `ru-vosk-*` voices (ВОЗ → "вэ о зэ")
    /// and English on ONNX Kokoro builds (FBI → "ef bee eye"). No effect on
    /// FluidAudio Kokoro, `macos-*` or non-English voices, which spell
    /// initialisms in their own G2P; those paths warn on stderr instead (#842).
    /// `<say-as interpret-as="characters">` in SSML remains honored.
    #[arg(long = "no-expand-abbrev", default_value_t = false)]
    pub no_expand_abbrev: bool,
}

/// Resolve `--format` / `--bitrate` / `--sample-rate` / `--out` into a
/// [`tts::OutputFormat`]. Priority: explicit flag > `--out` extension > Wav
/// default (preserves historical stdout-RIFF behaviour). See #223.
pub(crate) fn resolve_output_format(
    format: Option<&str>,
    bitrate: Option<i32>,
    sample_rate: Option<u32>,
    out: Option<&std::path::Path>,
) -> Result<tts::OutputFormat, String> {
    use std::str::FromStr;

    // #275 D10: source label fed to the dtrace probe at the bottom.
    let (mut chosen, source): (tts::OutputFormat, &'static str) = match (format, out) {
        (Some(f), _) => (tts::OutputFormat::from_str(f)?, "--format"),
        (None, Some(p)) => {
            let ext_fmt = p
                .extension()
                .and_then(|e| e.to_str())
                .and_then(tts::encode::format_from_extension);
            match ext_fmt {
                Some(fmt) => (fmt, "out-ext"),
                None => (tts::OutputFormat::default(), "default"),
            }
        }
        (None, None) => (tts::OutputFormat::default(), "default"),
    };

    if let tts::OutputFormat::OggOpus {
        bitrate: ref mut br,
        sample_rate: ref mut sr,
    } = chosen
    {
        if let Some(b) = bitrate {
            *br = b;
        }
        if let Some(r) = sample_rate {
            *sr = r;
        }
        // The encoder keeps its own check, but reaching it costs a full synthesis first (T1-5).
        if !tts::encode::OPUS_BITRATE_RANGE.contains(br) {
            return Err(format!("--bitrate must be 6000..=510000 bps, got {br}"));
        }
    } else if bitrate.is_some() || sample_rate.is_some() {
        return Err("--bitrate / --sample-rate only apply to --format ogg-opus".to_string());
    }

    crate::dtrace!("format::resolved chosen={chosen:?} source={source}");
    Ok(chosen)
}

fn list_kokoro_voices(_cache: &std::path::Path) -> Vec<String> {
    #[cfg(all(
        feature = "system_kokoro",
        target_os = "macos",
        target_arch = "aarch64"
    ))]
    {
        return tts::fluid_kokoro::available_voice_ids();
    }
    #[allow(unreachable_code)]
    {
        let dir = _cache.join("models/kokoro-82m/voices");
        std::fs::read_dir(&dir)
            .into_iter()
            .flatten()
            .filter_map(|e| e.ok())
            .filter_map(|e| {
                let p = e.path();
                if p.extension().and_then(|s| s.to_str()) != Some("bin") {
                    return None;
                }
                let stem = p.file_stem()?.to_string_lossy().into_owned();
                // Only prefixes the non-ANE resolve_voice arm accepts may be listed; an id the resolver rejects is #1168 again.
                let lang = match stem.chars().next()? {
                    'a' | 'b' => "en",
                    'e' => "es",
                    'f' => "fr",
                    'i' => "it",
                    'p' => "pt",
                    _ => return None,
                };
                Some(format!("{lang}-{stem}"))
            })
            .collect()
    }
}

fn list_vosk_ru_voices(cache: &std::path::Path) -> Vec<String> {
    // Vosk-TTS Russian is a single multi-speaker model — once installed, all
    // five baked-in speakers are available. Same gate as resolve_vosk_ru, so
    // partial installs don't advertise voices that fail at synthesis time.
    let dir = models::model_dir_at(models::ModelKind::VoskRu, cache);
    if !models::is_cached_in(models::ModelKind::VoskRu, &dir) {
        return Vec::new();
    }
    vec![
        "ru-vosk-f01".into(),
        "ru-vosk-f02".into(),
        "ru-vosk-f03".into(),
        "ru-vosk-m01".into(),
        "ru-vosk-m02".into(),
    ]
}

/// Map a TTS error to the documented exit code for `kesha say`: 1 = operational,
/// 2 = bad input, 4 = uncoded or unservable, 5 = text too long.
/// Keyed on the code, so the same code cannot exit differently per engine (T2-5).
fn exit_code_for_tts_err(e: &tts::TtsError) -> i32 {
    use crate::errors::ErrorCode as C;
    match e.code() {
        C::ModelMissing
        | C::ModelDownload
        | C::CacheCorrupt
        | C::ModelLoad
        | C::SidecarMissing
        | C::VoiceUnknown => 1,
        C::InvalidArg | C::SsmlInvalid | C::TextEmpty => 2,
        C::TextTooLong => 5,
        _ => 4,
    }
}

/// Read text from stdin, trimming surrounding whitespace.
fn read_stdin() -> Result<String, i32> {
    use std::io::Read;
    // Four bytes per allowed character: a pipe that streams past it is refused without waiting for EOF.
    const MAX_STDIN_BYTES: u64 = tts::MAX_TEXT_CHARS as u64 * 4 + 4;
    let mut bytes = Vec::new();
    if let Err(e) = std::io::stdin()
        .lock()
        .take(MAX_STDIN_BYTES + 1)
        .read_to_end(&mut bytes)
    {
        events::error(
            crate::errors::ErrorCode::Internal,
            format!("failed to read stdin: {e}"),
            None,
        );
        return Err(4);
    }
    if bytes.len() as u64 > MAX_STDIN_BYTES {
        let err = tts::TtsError::TextTooLong {
            max: tts::MAX_TEXT_CHARS,
            actual: String::from_utf8_lossy(&bytes).chars().count(),
        };
        events::error(err.code(), format!("{err}"), None);
        return Err(exit_code_for_tts_err(&err));
    }
    Ok(String::from_utf8_lossy(&bytes).trim().to_string())
}

/// Validate text length against TTS limits; returns the validated string or an
/// exit code on failure.
fn validate_text(text: String) -> Result<String, i32> {
    if text.is_empty() {
        let err = tts::TtsError::EmptyText;
        events::error(err.code(), format!("{err}"), None);
        return Err(exit_code_for_tts_err(&err));
    }
    let len = text.chars().count();
    if len > tts::MAX_TEXT_CHARS {
        let err = tts::TtsError::TextTooLong {
            max: tts::MAX_TEXT_CHARS,
            actual: len,
        };
        events::error(err.code(), format!("{err}"), None);
        return Err(exit_code_for_tts_err(&err));
    }
    Ok(text)
}

/// Resolve `--model` + `--voice-file` overrides or look up the voice by id.
/// Returns `(ResolvedVoice, exit_code_on_err)`.
fn resolve_voice(
    model: Option<PathBuf>,
    voice_file: Option<PathBuf>,
    voice_id: Option<&str>,
) -> Result<tts::voices::ResolvedVoice, i32> {
    match (model, voice_file) {
        (Some(model_path), Some(voice_path)) => Ok(tts::voices::ResolvedVoice::Kokoro {
            model_path,
            voice_path,
            espeak_lang: "en-us",
        }),
        (Some(_), None) | (None, Some(_)) => {
            events::error(
                crate::errors::ErrorCode::InvalidArg,
                "pass both --model and --voice-file or neither",
                None,
            );
            Err(2)
        }
        (None, None) => {
            let id = voice_id.unwrap_or(tts::voices::DEFAULT_VOICE_ID);
            let cache = models::cache_dir().map_err(|err| crate::errors::report(&err))?;
            tts::voices::resolve_voice(&cache, id).map_err(|err| crate::errors::report(&err))
        }
    }
}

/// Build the [`tts::EngineChoice`] from the resolved voice and playback rate.
fn engine_choice<'a>(
    resolved: &'a tts::voices::ResolvedVoice,
    voice_id: &'a str,
    rate: f32,
) -> tts::EngineChoice<'a> {
    match resolved {
        tts::voices::ResolvedVoice::Kokoro {
            model_path,
            voice_path,
            ..
        } => tts::EngineChoice::Kokoro {
            model_path,
            voice_path,
            speed: rate,
        },
        #[cfg(all(
            feature = "system_kokoro",
            target_os = "macos",
            target_arch = "aarch64"
        ))]
        tts::voices::ResolvedVoice::FluidKokoro { voice_id, .. } => {
            tts::EngineChoice::FluidKokoro {
                voice_id,
                speed: rate,
            }
        }
        tts::voices::ResolvedVoice::Vosk {
            model_dir,
            speaker_id,
        } => tts::EngineChoice::Vosk {
            voice_id,
            model_dir,
            speaker_id: *speaker_id,
            speed: rate,
        },
        #[cfg(all(feature = "system_tts", target_os = "macos"))]
        tts::voices::ResolvedVoice::AVSpeech { voice_id } => tts::EngineChoice::AVSpeech {
            voice_id,
            speed: rate,
        },
    }
}

/// A fifo destination streams, so it is never opened here: the open would block until a reader attaches.
#[cfg(unix)]
fn out_file_type_refusal(path: &std::path::Path) -> Option<Result<(), String>> {
    use std::os::unix::fs::FileTypeExt;
    let file_type = std::fs::metadata(path).map(|m| m.file_type());
    // `/dev/stdout` is the engine's stdout, which is the CLI's pipe, not the caller's (T1-15).
    let device = path.starts_with("/dev")
        || file_type
            .as_ref()
            .is_ok_and(|ft| ft.is_char_device() || ft.is_block_device());
    if device {
        return Some(Err(format!(
            "--out {} is a character device, where the audio would be discarded; \
             omit --out to write it to stdout",
            path.display()
        )));
    }
    file_type.is_ok_and(|ft| ft.is_fifo()).then_some(Ok(()))
}

#[cfg(not(unix))]
fn out_file_type_refusal(_path: &std::path::Path) -> Option<Result<(), String>> {
    None
}

/// Probe `--out` before synthesis pays for a path the caller mistyped, as `record::WavOutput::open` does (T1-4).
fn probe_out_path(path: &std::path::Path) -> Result<(), String> {
    if let Some(verdict) = out_file_type_refusal(path) {
        return verdict;
    }
    let existed = path.exists();
    std::fs::OpenOptions::new()
        .write(true)
        .create(true)
        .truncate(false)
        .open(path)
        .map_err(|err| format!("cannot write --out {}: {err}", path.display()))?;
    if !existed {
        let _ = std::fs::remove_file(path);
    }
    Ok(())
}

/// Write synthesized bytes to `--out` file or stdout.
fn write_output(out: Option<&std::path::Path>, bytes: &[u8]) -> Result<(), i32> {
    use std::io::Write;
    let result = match out {
        Some(p) => std::fs::write(p, bytes).map_err(|e| e.to_string()),
        None => std::io::stdout()
            .write_all(bytes)
            .map_err(|e| e.to_string()),
    };
    result.map_err(|msg| {
        events::error(
            crate::errors::ErrorCode::Internal,
            format!("write failed: {msg}"),
            None,
        );
        4
    })
}

pub fn run(a: SayArgs) -> i32 {
    if a.list_voices {
        let cache = match models::cache_dir() {
            Ok(c) => c,
            Err(err) => return crate::errors::report(&err),
        };
        let mut voice_ids: Vec<String> = list_kokoro_voices(&cache)
            .into_iter()
            .chain(list_vosk_ru_voices(&cache))
            .collect();
        // macos-* voices live in the OS, not the cache — enumerate them via
        // the AVSpeech helper (#141). Best-effort: if the helper is absent or
        // errors out, we still show Kokoro/Vosk voices.
        #[cfg(all(feature = "system_tts", target_os = "macos"))]
        voice_ids.extend(tts::avspeech::list_voices(None));
        voice_ids.sort();
        // Stdout is the list: a sentence there is a voice id to the MCP list_voices tool (#1168).
        if voice_ids.is_empty() {
            events::progress(None, "No voices installed. Run: kesha install --tts");
        }
        for id in voice_ids {
            println!("{id}");
        }
        return 0;
    }

    if a.stdin_loop {
        return say_loop::run();
    }

    let format = match resolve_output_format(
        a.format.as_deref(),
        a.bitrate,
        a.sample_rate,
        a.out.as_deref(),
    ) {
        Ok(f) => f,
        Err(msg) => {
            events::error(crate::errors::ErrorCode::InvalidArg, msg, None);
            return 2;
        }
    };

    if let Some(path) = a.out.as_deref() {
        if let Err(msg) = probe_out_path(path) {
            events::error(crate::errors::ErrorCode::InvalidArg, msg, None);
            return 2;
        }
    }

    if let Err(msg) = tts::say::validate_rate(a.rate) {
        events::error(crate::errors::ErrorCode::InvalidArg, msg, None);
        return 2;
    }

    let raw_text = match a.text {
        Some(s) => s,
        None => match read_stdin() {
            Ok(s) => s,
            Err(code) => return code,
        },
    };

    let text = match validate_text(raw_text) {
        Ok(t) => t,
        Err(code) => return code,
    };

    let resolved = match resolve_voice(a.model, a.voice_file, a.voice.as_deref()) {
        Ok(r) => r,
        Err(code) => return code,
    };

    let espeak_lang = a
        .lang
        .clone()
        .unwrap_or_else(|| resolved.espeak_lang().to_string());
    let engine = engine_choice(
        &resolved,
        a.voice.as_deref().unwrap_or(tts::voices::DEFAULT_VOICE_ID),
        a.rate,
    );

    let bytes = match tts::say(tts::SayOptions {
        text: &text,
        lang: &espeak_lang,
        engine,
        ssml: a.ssml,
        format,
        expand_abbrev: !a.no_expand_abbrev,
    }) {
        Ok(w) => w,
        Err(e) => {
            events::error(e.code(), format!("{e}"), None);
            return exit_code_for_tts_err(&e);
        }
    };

    match write_output(a.out.as_deref(), &bytes) {
        Ok(()) => 0,
        Err(code) => code,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::Path;

    #[test]
    fn explicit_format_wins_over_out_extension() {
        let fmt = resolve_output_format(Some("wav"), None, None, Some(Path::new("x.flac")))
            .expect("wav with flac out path");
        assert_eq!(fmt, tts::OutputFormat::Wav);
    }

    #[test]
    fn out_extension_selects_format() {
        for (path, expected) in [
            ("a.oga", tts::OutputFormat::ogg_opus_default()),
            ("a.ogg", tts::OutputFormat::ogg_opus_default()),
            ("a.flac", tts::OutputFormat::Flac),
            ("a.wav", tts::OutputFormat::Wav),
            ("a.txt", tts::OutputFormat::Wav),
        ] {
            let fmt = resolve_output_format(None, None, None, Some(Path::new(path)))
                .unwrap_or_else(|e| panic!("{path}: {e}"));
            assert_eq!(fmt, expected, "{path}");
        }
    }

    #[test]
    fn defaults_to_wav_without_hints() {
        assert_eq!(
            resolve_output_format(None, None, None, None).unwrap(),
            tts::OutputFormat::Wav
        );
    }

    #[test]
    fn opus_knobs_apply_to_ogg_opus() {
        let fmt = resolve_output_format(Some("ogg-opus"), Some(64_000), Some(48_000), None)
            .expect("explicit opus with knobs");
        assert_eq!(
            fmt,
            tts::OutputFormat::OggOpus {
                bitrate: 64_000,
                sample_rate: 48_000
            }
        );
    }

    #[test]
    fn opus_knobs_rejected_for_other_formats() {
        for (format, out) in [
            (Some("wav"), None),
            (Some("flac"), None),
            (None, Some(Path::new("a.flac"))),
        ] {
            let err = resolve_output_format(format, Some(32_000), None, out)
                .expect_err("bitrate must be rejected off the opus path");
            assert!(err.contains("only apply to --format ogg-opus"), "{err}");
        }
        let err = resolve_output_format(Some("wav"), None, Some(24_000), None)
            .expect_err("sample-rate must be rejected off the opus path");
        assert!(err.contains("only apply to --format ogg-opus"), "{err}");
    }

    #[test]
    fn opus_bitrate_range_is_enforced_before_synthesis() {
        for bad in [1, 5_999, 510_001] {
            let err = resolve_output_format(Some("ogg-opus"), Some(bad), None, None)
                .expect_err("bitrate outside the documented range must be refused");
            assert!(err.contains("--bitrate must be 6000..=510000 bps"), "{err}");
        }
        for ok in [6_000, 32_000, 510_000] {
            resolve_output_format(Some("ogg-opus"), Some(ok), None, None)
                .unwrap_or_else(|e| panic!("{ok}: {e}"));
        }
    }

    #[test]
    fn unknown_format_lists_supported_values() {
        let err = resolve_output_format(Some("mp3"), None, None, None).unwrap_err();
        assert!(err.contains("supported: wav, ogg-opus, flac"), "{err}");
    }

    #[test]
    fn tts_exit_codes_match_documented_contract() {
        assert_eq!(exit_code_for_tts_err(&tts::TtsError::EmptyText), 2);
        assert_eq!(
            exit_code_for_tts_err(&tts::TtsError::TextTooLong {
                max: 5000,
                actual: 5001
            }),
            5
        );
        assert_eq!(
            exit_code_for_tts_err(&tts::TtsError::SynthesisFailed("boom".into())),
            4
        );
        use crate::errors::ErrorCode as C;
        for (code, expected) in [
            (C::ModelMissing, 1),
            (C::ModelDownload, 1),
            (C::CacheCorrupt, 1),
            (C::ModelLoad, 1),
            (C::SidecarMissing, 1),
            (C::VoiceUnknown, 1),
            (C::InvalidArg, 2),
            (C::SsmlInvalid, 2),
            (C::TextEmpty, 2),
            (C::TextTooLong, 5),
            (C::ScriptUnsupported, 4),
            (C::SsmlUnsupported, 4),
            (C::Internal, 4),
        ] {
            assert_eq!(
                exit_code_for_tts_err(&tts::TtsError::Coded {
                    code,
                    message: "boom".into()
                }),
                expected,
                "{}",
                code.as_str()
            );
        }
    }
}
